neurips23/streaming/base_postgres.py

```diff
@@ -0,0 +1,274 @@
+import subprocess
+import time
+import numpy as np
+import psycopg
+import concurrent.futures
+
+from math import ceil
+from neurips23.streaming.base import BaseStreamingANN
+from pgvector.psycopg import register_vector
+
+
+PG_CONN_STR = "dbname=postgres user=postgres port=5432 host=localhost"
+
+
+def cursor_print_and_execute(cur, query):
+    print(query)
+    cur.execute(query)
+
+class BaseStreamingANNPostgres(BaseStreamingANN):
+    # Child classes should implement the following methods ..:
+    # - determine_index_op_class(self, metric)
+    # - determine_query_op(self, metric)
+    #
+    # .. as well as setting the following attributes in their __init__ methods before calling super().__init__:
+    # - self.name
+    # - self.pg_index_method
+    # - self.guc_prefix
+    def determine_index_op_class(self, metric):
+        raise NotImplementedError()
+
+    def determine_query_op(self, metric):
+        raise NotImplementedError()
+
+    def __init__(self, metric, index_params):
+        self.n_insert_conns = index_params.get("insert_conns")
+        if self.n_insert_conns == None:
+            raise Exception('Missing parameter insert_conns')
+
+        # save it for later use in __str__()
+        self._index_params = index_params
+
+        # we'll initialize the connections later in set_query_arguments() per "query-arguments" set
+        self.conns = []
+
+        self.index_build_params = {k: v for k, v in index_params.items() if k != "insert_conns"}
+
+        self.ind_op_class = self.determine_index_op_class(metric)
+
+        self.search_query = f"SELECT id FROM test_tbl ORDER BY vec_col {self.determine_query_op(metric)} %b LIMIT %b"
+
+        start_database_result = subprocess.run(['bash', '/home/app/start_database.sh'], capture_output=True, text=True)
+        if start_database_result.returncode != 0:
+            raise Exception(f'Failed to start the database: {start_database_result.stderr}')
+
+    def setup(self, dtype, max_pts, ndim):
+        if dtype != 'float32':
+            raise Exception('Invalid data type')
+
+        index_build_params_clause = ""
+        if self.index_build_params:
+            index_build_params_clause = "WITH ("
+            first = True
+            for k, v in self.index_build_params.items():
+                if not first:
+                    index_build_params_clause += ", "
+
+                first = False
+                index_build_params_clause += f"{k} = {v}"
+
+            index_build_params_clause += ")"
+
+        # create the table and index by using a temporary connection
+        with psycopg.connect(PG_CONN_STR, autocommit=True) as conn:
+            with conn.cursor() as cur:
+                cursor_print_and_execute(cur, f"CREATE TABLE test_tbl (id bigint, vec_col vector({ndim}))")
+                cursor_print_and_execute(cur, f"CREATE INDEX vec_col_idx ON test_tbl USING {self.pg_index_method} (vec_col {self.ind_op_class}) {index_build_params_clause}")
```

Comment:
To stabilize the measurements, should we disable autovacuum when creating the table since we explicitly vacuum the table at some point? @orhankislal?